### PR TITLE
QML Pt. 2: Add demo skin

### PR DESCRIPTION
**Pull request by Holzhaus**
Based on #3891 and #3894. This is just the WIP QML demo skin, no C++ changes.

### CODE REVIEW

File: res/skins/QMLDemo/Deck.qml
Comment (Swiftb0y) [inline res/skins/QMLDemo/Deck.qml:35]:
```suggestion
                        anchors.verticalCenter: parent.verticalCenter
                        anchors.left: parent.left
```
can you try this out? not sure if it has the same result though...

File: res/skins/QMLDemo/Deck.qml
Comment (Swiftb0y) [inline res/skins/QMLDemo/Deck.qml:36]: Not sure if hardcoding the 1:1 aspect ration here is a good idea, but we can keep it like this for the sake of simplicity I guess. I'd prefer if this was documented in an inline comment though.

File: res/skins/QMLDemo/Deck.qml
Comment (Swiftb0y) [inline res/skins/QMLDemo/Deck.qml:7]: This is already quite big, I'd prefer to split everything into its own file if possible. At least Items which are already decoupled. Accessing properties from the deck itself could either be done by making use of [QMLs dynamic scoping](https://doc.qt.io/qt-5/qtqml-documents-scope.html) or by using (alias) properties.

File: res/skins/QMLDemo/EqKnob.qml
Comment (Swiftb0y) [inline res/skins/QMLDemo/EqKnob.qml:33]: You could use a Component to keep the code a bit more DRY:
```suggestion
        knobImage: Component {
            property alias source: image.source
            Image: {
              id: image
              width: 35
              height: 30
            }
        }
        background: knobImage {
            source: "../LateNight/palemoon/knobs/knob_bg_master.svg"
        }

        foreground: knobImage {
            source: "../LateNight/palemoon/knobs/knob_indicator_regular_red.svg"
        }
```
Its overkill in this case, but a handy technique to create components inline if they don't deserve their own file.

**Review by Swiftb0y — COMMENTED** [review]
Took a quick look at some of QML code and made some comments about what I consider best practice from my limited research on QML. I haven't had the opportunity to run the demo in practice yet.

Styling is still a topic to be discussed. QtQuick primarily seems to support styling by using visual objects to delegate the drawing to the caller/instatiator. We might want to consider using the same pattern for QML components we ship with mixxx.

File: res/skins/QMLDemo/Deck.qml
Comment (Holzhaus) [reply res/skins/QMLDemo/Deck.qml:35]: No, this centers the item, but it won't show up because it has no height. The current code sets the height to that of the parent.

File: res/skins/QMLDemo/Deck.qml
Comment (Holzhaus) [reply res/skins/QMLDemo/Deck.qml:36]: FYI I tried to make a separation between skin-local QML and reusable QML that may be reused by multiple skins, similar to controller mappings vs. midi-components. The latter is in the Mixxx QML module (subdirectory).

File: res/skins/QMLDemo/Deck.qml
Comment (Holzhaus) [reply res/skins/QMLDemo/Deck.qml:7]: Hmm, I'd rather not make too many changes to not cause too many conflicts with the follow-up QML PRs. This is just a demo skin anyway, so we can refactor later :D It's more important that the reusable components in the `Mixxx` subdirectory are well-structured.

File: res/skins/QMLDemo/EqKnob.qml
Comment (Holzhaus) [reply res/skins/QMLDemo/EqKnob.qml:33]: Good point, thanks. Does this work in all QtQuick versions? I have Qt 5.15 and don't know if the current version even works on Qt 5.12.

File: res/skins/QMLDemo/Deck.qml
Comment (Swiftb0y) [reply res/skins/QMLDemo/Deck.qml:35]: Ok good to know, so the anchor doubles as a responsive layout guide.

File: res/skins/QMLDemo/Deck.qml
Comment (Swiftb0y) [reply res/skins/QMLDemo/Deck.qml:7]: As long as it is a Demo skin thats fine, but the first QML skin should be well polished and using QML best practices. Otherwise others will copy that mediocre code and we end up with a similar situation as with the mappings.

File: res/skins/QMLDemo/EqKnob.qml
Comment (Swiftb0y) [reply res/skins/QMLDemo/EqKnob.qml:33]: I'm pretty pretty sure it does work on all versions.
Making the alias properties required would make sense as well, but I'm not sure if that has implications for the Qt version.

File: res/skins/QMLDemo/Deck.qml
Comment (Holzhaus) [reply res/skins/QMLDemo/Deck.qml:7]: Yes. I hope that once the demo skin is merged, other people start working on it and improve it, until we end up with a nice, polished skin.

**Comment by Holzhaus** [comment]
I also added a pure-QML implementation of spinnies. Ready now.

**Comment by Holzhaus** [comment]
I rebased on latest `main`. Ready to test and review. I do not intend to do further changes to this PR unless you find issues.

I'm running Qt 5.15, so it's possible that some QtQuick components are not present in Qt 5.12. Maybe someone can check that.

**Comment by Holzhaus** [comment]
OK, apparently I mustn't use any QML modules with a version greater than `2.12` (for Qt 5.12). I'll check if and where I do that.

**Comment by Swiftb0y** [comment]
> OK, apparently I mustn't use any QML modules with a version greater than 2.12 (for Qt 5.12). I'll check if and where I do that.

Afaik QML Modules follow the Qt version unless they have not been changed. If they haven't been changed they stay on the version they last have been changed in. In general, we can only use QML modules with a version lower or equal than the minimum Qt version we support.

**Review by uklotzde — APPROVED** [review]
@Swiftb0y Please merge when you think this is ready.

**Comment by Holzhaus** [comment]
I downgraded all module versions to the Qt 5.12 versions.

File: res/skins/QMLDemo/EqKnob.qml
Comment (Holzhaus) [reply res/skins/QMLDemo/EqKnob.qml:33]: > Making the alias properties required would make sense as well

Nope, not possible. The normal properties you use one of these:

    [default] [required] [readonly] property <propertyType> <propertyName>
    [default] property <propertyType> <propertyName> : <value>

But for alias properties the signature is:

  [default] property alias <name>: <alias reference>

So you can't make them required.

https://doc.qt.io/qt-5/qtqml-syntax-objectattributes.html#property-aliases

Comment (Swiftb0y) [reply res/skins/QMLDemo/EqKnob.qml:33]: Ok, good to know, so I guess alias properties would get required automatically when they refer to a required resource?

Comment (Holzhaus) [reply res/skins/QMLDemo/EqKnob.qml:33]: I guess so.

**Comment by Swiftb0y** [comment]
I'm building the branch right now to try it out. I'll merge if I don't notice any issues during testing.

File: res/skins/QMLDemo/Mixxx/Controls/Spinny.qml
Comment (Swiftb0y) [inline res/skins/QMLDemo/Mixxx/Controls/Spinny.qml:55]: I'm a little bit worried how often this will get called. it would be ideal if it was just once per frame but can we guarantee that?

**Review by Swiftb0y — COMMENTED** [review]
otherwise LGTM

File: res/skins/QMLDemo/Mixxx/Controls/Spinny.qml
Comment (Holzhaus) [reply res/skins/QMLDemo/Mixxx/Controls/Spinny.qml:55]: I'll check if we can improve it by using a declarative binding that defines a relationship. I guess QML is then smart enough to figure it out itself.

Comment (Holzhaus) [reply res/skins/QMLDemo/Mixxx/Controls/Spinny.qml]: Done.

**Review by Swiftb0y — APPROVED** [review]
Thank you for your patience

**Pull request merged**
